drivers/audio/wm8960.c

```diff
@@ -0,0 +1,752 @@
+		break;
+	case AUDIO_ROUTE_PLAYBACK:
+		/*
+		 * Data route I2S_IN-> DAC-> HP
+		 * Left DAC to left output mixer,
+		 * LINPUT3 left output mixer volume = 0dB
+		 */
+		codec_write_reg(dev, WM8960_LOUTMIX, 0x100);
+
+		/*
+		 * Right DAC to right output mixer,
+		 * RINPUT3 right output mixer volume = 0dB
+		 */
+		codec_write_reg(dev, WM8960_ROUTMIX, 0x100);
+		codec_write_reg(dev, WM8960_POWER3, 0x0C);
+		/* Set power for DAC */
+		codec_write_reg(dev, kWM8960_ModuleDAC, true);
+		codec_write_reg(dev, kWM8960_ModuleOMIX, true);
+		codec_write_reg(dev, kWM8960_ModuleLineOut, true);
+		break;
+	case AUDIO_ROUTE_PLAYBACK_CAPTURE:
+		/*
+		 * Left DAC to left output mixer,
+		 * LINPUT3 left output mixer volume = 0dB
+		 */
+		codec_write_reg(dev, WM8960_LOUTMIX, 0x100);
+
+		/*
+		 * Right DAC to right output mixer,
+		 * RINPUT3 right output mixer volume = 0dB
+		 */
+		codec_write_reg(dev, WM8960_ROUTMIX, 0x100);
+		codec_write_reg(dev, WM8960_POWER3, 0x3C);
+		codec_write_reg(dev, kWM8960_ModuleDAC, true);
+		codec_write_reg(dev, kWM8960_ModuleADC, true);
+		codec_write_reg(dev, kWM8960_ModuleLineIn, true);
+		codec_write_reg(dev, kWM8960_ModuleOMIX, true);
+		codec_write_reg(dev, kWM8960_ModuleLineOut, true);
+		break;
+	case AUDIO_ROUTE_CAPTURE:
+		/*
+		 * LINE_IN->ADC->I2S_OUT
+		 * Left and right input boost, LIN3BOOST and RIN3BOOST = 0dB
+		 */
+		codec_write_reg(dev, WM8960_POWER3, 0x30);
+		/* Power up ADC and AIN */
+		codec_write_reg(dev, kWM8960_ModuleLineIn, true);
+		codec_write_reg(dev, kWM8960_ModuleADC, true);
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int codec_protol_config(const struct device *dev,
+			       audio_dai_type_t dai_type)
+{
+	enum _wm8960_protocol protocol;
+
+	switch (dai_type) {
+	case AUDIO_DAI_TYPE_I2S:
+		protocol = kWM8960_BusI2S;
+		break;
+	default:
+		return -EINVAL;
+	}
+	codec_update_reg(dev, WM8960_IFACE1,
+			 WM8960_IFACE1_FORMAT_MASK | WM8960_IFACE1_LRP_MASK,
+			 protocol);
+	LOG_DBG("protocol is 0x%x", protocol);
+	return 0;
+}
+
+static int codec_configure(const struct device *dev,
+			   struct audio_codec_cfg *cfg)
+{
+	int ret;
+
+	if (cfg->dai_type != AUDIO_DAI_TYPE_I2S) {
+		LOG_ERR("dai_type must be AUDIO_DAI_TYPE_I2S");
+		return -EINVAL;
+	}
+
+	codec_soft_reset(dev);
+
+	/*
+	 * VMID=50K, Enable VREF, AINL, AINR, ADCL and ADCR
+	 * I2S_IN (bit 0), I2S_OUT (bit 1), DAP (bit 4),
+	 * DAC (bit 5), ADC (bit 6) are powered on
+	 */
+	codec_write_reg(dev, WM8960_POWER1, 0xFE);
+
+	/*
+	 * Enable DACL, DACR, LOUT1, ROUT1, PLL down
+	 */
+	codec_write_reg(dev, WM8960_POWER2, 0x1E0);
+	/*
+	 * Enable left and right channel input PGA, left and right output mixer
+	 */
+	codec_write_reg(dev, WM8960_POWER3, 0x3C);
+	/* ADC and DAC uses same clock */
+	codec_write_reg(dev, WM8960_IFACE2, 0x40);
+	ret = code_set_data_route(dev, cfg->dai_route);
+	if (ret == 0) {
+		ret = codec_protol_config(dev, cfg->dai_type);
+		LOG_DBG("config protol done");
```

Comment:
typo, should be "protocol"

Reply:
The return value be checked and return an error if not zero.

Reply:
WM8960 may not support all configs, it will retrun -EINVA.

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/37335#discussion_r694112487

thanks correct it.